features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/support/ParameterSubstitutingMonitor.java

```diff
@@ -0,0 +1,177 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.poller.monitors.support;
+
+import java.beans.BeanInfo;
+import java.beans.Introspector;
+import java.beans.IntrospectionException;
+import java.beans.PropertyDescriptor;
+import java.io.UnsupportedEncodingException;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+
+import org.opennms.core.spring.BeanUtils;
+import org.opennms.core.utils.ParameterMap;
+import org.opennms.netmgt.dao.api.NodeDao;
+import org.opennms.netmgt.model.OnmsNode;
+import org.opennms.netmgt.model.OnmsAssetRecord;
+import org.opennms.netmgt.poller.MonitoredService;
+import org.opennms.netmgt.poller.support.AbstractServiceMonitor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.beans.BeanWrapper;
+import org.springframework.beans.BeanWrapperImpl;
+
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+/**
+ * {@Inheritdoc}
+ * Uses {@link org.opennms.netmgt.poller.ServiceMonitor#getRuntimeAttributes(MonitoredService svc, Map<String, Object> parameters)}
+ * to add parameters with substitutions of placeholders in existing parameters.
+ * The new parameter keys will be prepended with <code>'subbed-'</code>.
+ *
+ * @author <A HREF="dschlenk@convergeone.com">David Schlenk</A>
+ */
+public abstract class ParameterSubstitutingMonitor extends AbstractServiceMonitor {
+
+    public static final Logger LOG = LoggerFactory.getLogger(ParameterSubstitutingMonitor.class);
+
+    private static Map<String, Pattern> subPatterns = new HashMap<>();
+    private static Pattern m_subPat;
+    private static final Supplier<NodeDao> nodeDao = Suppliers.memoize(() -> BeanUtils.getBean("daoContext", "nodeDao", NodeDao.class));
+    static {
+        StringBuilder patternBuilder = new StringBuilder();
+        patternBuilder.append(".*[{](ipAddr(?:ess)?|nodeId|nodeLabel|foreignId|foreignSource");
+        try {
+            BeanInfo info = Introspector.getBeanInfo(OnmsAssetRecord.class);
+            PropertyDescriptor[] pds = info.getPropertyDescriptors();
+            for (PropertyDescriptor pd : pds) {
+                patternBuilder.append("|").append(pd.getName());
+                subPatterns.put(pd.getName(), Pattern.compile("(.*)[{]" + pd.getName() + "[}](.*)"));
+            }
+        } catch (IntrospectionException ie) {
+            ie.printStackTrace();
```

Comment:
Can we use the logger here and/or rethrow the exception instead of printing the stack trace?